symengine/parser/sbml/sbml_parser.cpp

```diff
@@ -0,0 +1,386 @@
+                                                      const RCP<const Basic> &);
+typedef RCP<const Basic> (*vector_arg_func)(const vec_basic &);
+
+// cast overloaded functions below to single_arg, double_arg before they can
+// be used in the map
+static single_arg_func single_casted_log = log;
+
+static double_arg_boolean_func double_casted_Eq = Eq;
+
+static vector_arg_func vector_casted_add = add;
+static vector_arg_func vector_casted_mul = mul;
+
+static RCP<const Basic> sqr(const RCP<const Basic> &x)
+{
+    return pow(x, integer(2));
+}
+
+static RCP<const Basic> sbml_log(const RCP<const Basic> &b,
+                                 const RCP<const Basic> &x)
+{
+    return log(x, b);
+}
+
+static RCP<const Basic> log10(const RCP<const Basic> &x)
+{
+    return log(x, integer(10));
+}
+
+static RCP<const Basic> root(const RCP<const Basic> &b,
+                             const RCP<const Basic> &x)
+{
+    return pow(x, div(integer(1), b));
+}
+
+static RCP<const Basic> fact(const RCP<const Basic> &x)
+{
+    return gamma(add(x, integer(1)));
+}
+
+static RCP<const Boolean> vec_ge(const vec_basic &x)
+{
+    set_boolean b;
+    for (std::size_t i = 0; i + 1 < x.size(); ++i) {
+        b.insert(Ge(x[i], x[i + 1]));
+    }
+    return logical_and(b);
+}
+
+static RCP<const Boolean> vec_gt(const vec_basic &x)
+{
+    set_boolean b;
+    for (std::size_t i = 0; i + 1 < x.size(); ++i) {
+        b.insert(Gt(x[i], x[i + 1]));
+    }
+    return logical_and(b);
+}
+
+static RCP<const Boolean> vec_le(const vec_basic &x)
+{
+    set_boolean b;
+    for (std::size_t i = 0; i + 1 < x.size(); ++i) {
+        b.insert(Le(x[i], x[i + 1]));
+    }
+    return logical_and(b);
+}
+
+static RCP<const Boolean> vec_lt(const vec_basic &x)
+{
+    set_boolean b;
+    for (std::size_t i = 0; i + 1 < x.size(); ++i) {
+        b.insert(Lt(x[i], x[i + 1]));
+    }
+    return logical_and(b);
+}
+
+RCP<const Basic> SbmlParser::modulo(const RCP<const Basic> &a,
+                                    const RCP<const Basic> &b)
+{
+    return piecewise({{sub(a, mul(b, ceiling(div(a, b)))),
+                       logical_xor({Lt(a, zero), Lt(b, zero)})},
+                      {sub(a, mul(b, floor(div(a, b)))), boolTrue}});
+}
+
+RCP<const Basic> SbmlParser::functionify(const std::string &name)
+{
+    const static std::map<const std::string, const RCP<const Basic>>
+        zero_arg_functions = {{"plus", integer(0)},
+                              {"times", integer(1)},
+                              {"and", boolTrue},
+                              {"or", boolFalse},
+                              {"xor", boolFalse}};
+    std::string lname = lowercase(name);
+    auto l = zero_arg_functions.find(lname);
+    if (l != zero_arg_functions.end()) {
+        return l->second;
+    }
+    throw ParseError("Parsing Unsuccessful: Function '" + name
+                     + "' has no arguments");
+}
+
+RCP<const Basic> SbmlParser::functionify(const std::string &name,
```

Comment:
We can reduce the code duplication here using the same trick as `init_sbml_printer_names`. What do you think?

Reply:
I tried this here: https://github.com/symengine/symengine/pull/1785/commits/a159dcd1464c8cc7847065177e6fa64918790c3b 

Is that what you meant? If so, it does reduce duplication, but it might also require more maintenance - e.g. if a new function is added to the main parser that is not supported by sbml, it would also have to be explicitly removed from the sbml parser. Either way is fine by me though.

Reply:
Having to remember to remove something in one end when something was added in another sounds like an extra maintenance burden. Don't we want to avoid that?

Reply:
Yes, looks good to me.

> if a new function is added to the main parser that is not supported by sbml, it would also have to be explicitly removed from the sbml parser

Is it necessary to remove them though? Would it be bad if they were kept?

Reply:
> Yes, looks good to me.
> 
> > if a new function is added to the main parser that is not supported by sbml, it would also have to be explicitly removed from the sbml parser
> 
> Is it necessary to remove them though? Would it be bad if they were kept?

- it could mean invalid sbml infix gets parsed as valid
  - but it would give an error when printing: so that avoids the invalid sbml from being exported somewhere else
- if a function is added to the main parser with the same name as an existing sbml one it would take precedence (since `insert` is a no-op if the key already exists), which could potentially change the meaning of this function in the sbml parser
  - but this would then cause a sbml_parser test to fail

tldr; no, it's probably fine to keep them

Reply:
I guess the question is what should the sbml parser/printer do with math that the main parser+symengine understands, but that is not part of sbml?

1. Raise an error
  - need to define all valid functions -> code duplication
  - or inherit default functions & remove invalid ones -> maintenance burden when defaults change
2. Fallback to the default printer/parser behaviour
  - no code duplication or maintenance issues

Currently we do `1.` when parsing sbml, and `2.` when printing sbml

- As a user I would prefer option `1.` for both printing & parsing: raising an error is much more helpful than silently parsing/producing invalid infix
- But the current setup is also fine, since the user can just re-parse the printed output to check it is valid
- As a maintainer I would probably prefer to implement this by defining all valid functions (i.e. duplication rather than maintenance) but either way is fine by me

What do you think?

Reply:
I also think the parser should throw on illegal sbml. Also the printer should throw for things that cannot be supported or isn't currenty supported in sbml. Some cases I guess could be supported with workarounds if needed. For example we have a printer for Fortran code in Pharmpy that converts `sign(x)` to `ABS(X) / X` (if x is known to be positive) or similar.

Reply:
@isuruf what are your thoughts on this?

Reply:
I'm okay with either.

Reply:
ok, I'll keep the current behaviour for now then (sbml parser throws on invalid sbml, sbml printer falls back to default printer), and leave improving the printer to throw on invalid sbml for a future PR.

I've reverted to defining all sbml parser functions to avoid having to update the sbml parser when new functions are added to the main parser (https://github.com/symengine/symengine/pull/1785/commits/4054db5bafce480dc825ad113c3109223fc49c8e)